libs/agent-sm/agent-policy/src/main/java/org/opensearch/secure_sm/policy/PolicyFile.java

```diff
@@ -308,10 +316,17 @@ public String toString() {
 
     private static class PolicyInfo {
         final List<PolicyEntry> policyEntries;
+        private final Map<ProtectionDomain, PermissionCollection> pdMapping;
 
         PolicyInfo() {
-            policyEntries = new ArrayList<>();
+            policyEntries = Collections.synchronizedList(new ArrayList<PolicyEntry>());
+            pdMapping = new ConcurrentHashMap<>();
         }
+
+        public PermissionCollection getOrCompute(ProtectionDomain pd, Supplier<PermissionCollection> computeFn) {
```

Comment:
Sorry, super nitpick here, but if you make the second parameter here `Function<ProtectionDomain, PermissionCollection> computeFn` then on line 191 you can just pass `this::getPermissions` and it's a little bit cleaner.